nexus/src/external_api/console_api.rs

```diff
@@ -718,115 +731,147 @@ fn with_gz_ext(path: &Utf8Path) -> Utf8PathBuf {
-        .join("assets");
+        .as_deref()
+        .ok_or_else(|| not_found("static_dir undefined"))?;
 
-    let request = &rqctx.request;
-    let accept_encoding = request.headers().get(http::header::ACCEPT_ENCODING);
-    let accept_gz = accept_encoding.map_or(false, |val| {
-        val.to_str().map_or(false, |s| s.contains("gzip"))
-    });
+    // Bail unless the extension is allowed
+    let content_type = ALLOWED_EXTENSIONS
+        .get(path.extension().ok_or_else(|| {
+            not_found("requested file does not have extension, not allowed")
+        })?)
+        .ok_or_else(|| not_found("file extension not allowed"))?;
+
+    let mut resp = Response::builder()
+        .status(StatusCode::OK)
+        .header(http::header::CONTENT_TYPE, content_type)
+        .header(http::header::CACHE_CONTROL, cache_control);
+    for (k, v) in WEB_SECURITY_HEADERS {
+        resp = resp.header(k, v);
+    }
 
     // If req accepts gzip and we have a gzipped version, serve that. Otherwise
     // fall back to non-gz. If neither file found, bubble up 404.
-    let (path_to_read, set_content_encoding_gzip) =
-        match accept_gz.then(|| find_file(&with_gz_ext(&path), &assets_dir)) {
-            Some(Ok(gzipped_path)) => (gzipped_path, true),
-            _ => (find_file(&path, &assets_dir)?, false),
-        };
+    let request = &rqctx.request;
+    let accept_encoding = request
+        .headers()
+        .get(http::header::ACCEPT_ENCODING)
+        .and_then(|v| v.to_str().ok())
+        .unwrap_or_default();
+    let path_to_read = match accept_gz(accept_encoding)
+        .then(|| find_file(&with_gz_ext(&path), static_dir))
+    {
+        Some(Ok(gzipped_path)) => {
+            resp = resp
+                .header(http::header::CONTENT_ENCODING, CONTENT_ENCODING_GZIP);
+            gzipped_path
+        }
+        _ => find_file(&path, static_dir)?,
+    };
 
-    // File read is the same regardless of gzip
-    let file_contents = tokio::fs::read(&path_to_read).await.map_err(|e| {
+    let file = File::open(&path_to_read).await.map_err(|e| {
         not_found(&format!("accessing {:?}: {:#}", path_to_read, e))
     })?;
+    let metadata = file.metadata().await.map_err(|e| {
+        not_found(&format!("accessing {:?}: {:#}", path_to_read, e))
+    })?;
+    resp = resp.header(http::header::CONTENT_LENGTH, metadata.len());
 
-    // Derive the MIME type from the file name (can't use path_to_read because
-    // it might end with .gz)
-    let content_type = path.file_name().map_or("text/plain", |f| {
-        mime_guess::from_path(f).first_raw().unwrap_or("text/plain")
-    });
-
-    let mut resp = Response::builder()
-        .status(StatusCode::OK)
-        .header(http::header::CONTENT_TYPE, content_type)
-        .header(http::header::CACHE_CONTROL, "max-age=31536000, immutable"); // 1 year
+    let stream = FramedRead::new(file, BytesCodec::new());
+    let body = Body::wrap_stream(stream);
```

Comment:
Does this mean we don't have to read the entire file into memory at once?

Reply:
Yep. That may have been leading to some request latency with the larger bundles, although not sure if it actually was in practice.

Reply:
Seems like probably not (this is colo, dogfood looks a bit different due to VPN I assume), but might as well do it the right way.

<img width="454" alt="image" src="https://github.com/oxidecomputer/omicron/assets/3612203/f2692a72-e023-40f8-a177-883957de21eb">